### PR TITLE
Fix the issue of server not registering state model factory before connecting the Helix manager

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
State model factory should be registered before connecting the Helix manager, while Helix property store is not available before connecting the Helix manager.
In order to solve this problem, passing Helix manager to the server instance and starting the server instance as a pre-connect callback in Helix manager.
Removed ServerBuilder to simplify the logic.

### CODE REVIEW

File: pinot-server/src/main/java/org/apache/pinot/server/starter/ServerInstance.java [hunk 1]
```diff
@@ -38,36 +45,64 @@
...
-  public void init(ServerConf serverConf, ZkHelixPropertyStore<ZNRecord> propertyStore)
+  public ServerInstance(ServerConf serverConf, HelixManager helixManager)
       throws Exception {
     LOGGER.info("Initializing server instance");
 
-    ServerBuilder serverBuilder = new ServerBuilder(serverConf, propertyStore);
-    _serverMetrics = serverBuilder.getServerMetrics();
-    _instanceDataManager = serverBuilder.buildInstanceDataManager();
-    _queryExecutor = serverBuilder.buildQueryExecutor(_instanceDataManager);
+    LOGGER.info("Initializing server metrics");
```
Comment (mayankshriv) [inline pinot-server/src/main/java/org/apache/pinot/server/starter/ServerInstance.java]: Nit: int()?

File: pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixServerStarter.java [hunk 2]
```diff
@@ -121,35 +120,33 @@ public HelixServerStarter(String helixClusterName, String zkServer, Configuratio
...
+    _serverInstance = new ServerInstance(serverInstanceConfig, _helixManager);
+    InstanceDataManager instanceDataManager = _serverInstance.getInstanceDataManager();
+    SegmentFetcherAndLoader fetcherAndLoader = new SegmentFetcherAndLoader(_serverConf, instanceDataManager);
     StateModelFactory<?> stateModelFactory =
-        new SegmentOnlineOfflineStateModelFactory(_instanceId, _serverInstance.getInstanceDataManager(),
-            fetcherAndLoader, propertyStore);
+        new SegmentOnlineOfflineStateModelFactory(_instanceId, instanceDataManager, fetcherAndLoader);
     _helixManager.getStateMachineEngine()
         .registerStateModelFactory(SegmentOnlineOfflineStateModelFactory.getStateModelName(), stateModelFactory);
+    // Start the server instance after connecting to the ZK but before processing state transitions
```
Comment (mayankshriv) [inline pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixServerStarter.java]: May be also elaborate the comment on why this is needed.

File: pinot-server/src/main/java/org/apache/pinot/server/starter/ServerInstance.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-server/src/main/java/org/apache/pinot/server/starter/ServerInstance.java]: I prefer constructor over init() as we can make member variables final and immutable

File: pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixServerStarter.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixServerStarter.java]: Done

**Pull request merged**